### PR TITLE
Notify the congestion controller of losses first

### LINKED ISSUE #3495: Loss-detection happens after CC-increase, leading to higher congestion window
In the pseudo-code in appendix A.6, the call to ```DetectLostPackets``` is after ```OnPacketAcked```, while the call to ```ProcessECN``` is before that:

```
  // Process ECN information if present.
  if (ACK frame contains ECN information):
      ProcessECN(ack, pn_space)

  for acked_packet in newly_acked_packets:
    OnPacketAcked(acked_packet.packet_number, pn_space)

  DetectLostPackets(pn_space)
```

This means that if I get an ACK while in slow-start that indicates a packet-loss, the stack will first call ```OnPacketAcked``` (which ends up increasing the congestion-window) and only later enter recovery through ```DetectLostPackets```. Meaning, the congestion-window will actually be bigger than the flight-size that caused the packet-loss.

E.g., if cwnd is 10 while in slow-start and 10 packets were sent. If I get an ACK for 9 packets, indicating a hole, the congestion-window will first increase to 19 and then enter congestion-avoidance and reduce the congestion-window to 13 (assuming Cubic). Thus, the congestion-window is effectively bigger than 10 while it should be smaller than 10 (the flight-size that caused the packet-loss).

Shouldn't we process the loss before ```OnPacketAcked```, the same way ECN is being processed before that?

### PR DESCRIPTION
**Pull request by ianswett**
Fixes #3495 and #3288 by notifying the congestion controller of lost packets prior to notifying it of acknowledged packets and passing all lost packets into InPersistentCongestion().

The diff is a bit large because I also inlined OnPacketAcked and renamed OnPacketAckedCC to OnPacketAcked.

### CODE REVIEW

**Comment by ghedo** [comment]
Thanks. I think `AreAllPacketsLost` will still need some clarifications but I haven't been able to find a nice way to word that, so for now this is good.

**Comment by cpaasch** [comment]
Looks good to me wrt to issue #3495

**Comment by ianswett** [comment]
@martinduke  I think I addressed your comment by commenting that DetectNewlyAckedPackets is responsible for removing packets from sent_packets.  This aligns with the API for DetectLostPackets.  I also changed OnPacketAcked to OnPacketsAcked.

"// Determine which packets have been acknowledged and remove
  // them from sent_packets.
  newly_acked_packets = DetectNewlyAckedPackets(ack, pn_space)"

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -1265,7 +1244,8 @@ OnLossDetectionTimeout():
     GetEarliestTimeAndSpace(loss_time)
   if (earliest_loss_time != 0):
     // Time threshold loss Detection
-    DetectLostPackets(pn_space)
+    lost_packets = DetectLostPackets(pn_space)
+    OnPacketsLost(lost_packets)
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
    if (!lost_packets.empty()):
      OnPacketsLost(lost_packets)
```

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -1517,7 +1494,7 @@ Invoked from DetectLostPackets when packets are deemed lost.
      CongestionEvent(largest_lost_packet.time_sent)
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: How about also changing this to the following, and removing the line before this (last() is a language-specific method, so I'm suggesting largest())?
```suggestion
     CongestionEvent(lost_packets.largest().time_sent)
```

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -1108,14 +1108,16 @@ OnAckReceived(ack, pn_space):
     largest_acked_packet[pn_space] =
         max(largest_acked_packet[pn_space], ack.largest_acked)
 
+  // Determine which packets have been acknowledged and remove
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
  // DetectNewlyAckedPackets finds packets have been newly acknowledged and removes
```

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -1108,14 +1108,16 @@ OnAckReceived(ack, pn_space):
...
+  // them from sent_packets.
+  newly_acked_packets = DetectNewlyAckedPackets(ack, pn_space)
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
  newly_acked_packets = DetectAndRemoveAckedPackets(ack, pn_space)
```
here and below

File: draft-ietf-quic-recovery.md [hunk 5]
```diff
@@ -1128,13 +1130,12 @@ OnAckReceived(ack, pn_space):
   if (ACK frame contains ECN information):
       ProcessECN(ack, pn_space)
 
-  for acked_packet in newly_acked_packets:
-    OnPacketAcked(acked_packet.packet_number, pn_space)
-
-  DetectLostPackets(pn_space)
+  lost_packets = DetectLostPackets(pn_space)
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
  lost_packets = DetectAndRemoveLostPackets(pn_space)
```
here and below

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -1434,26 +1411,27 @@ increases bytes_in_flight.
 ## On Packet Acknowledgement
 
 Invoked from loss detection's OnPacketAcked and is supplied with the
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
Invoked from loss detection's OnAckReceived and is supplied with the
```

File: draft-ietf-quic-recovery.md [hunk 7]
```diff
@@ -1499,15 +1477,14 @@ Invoked when an ACK frame with an ECN section is received from the peer.
 Invoked from DetectLostPackets when packets are deemed lost.
 
 ~~~
-   InPersistentCongestion(largest_lost_packet):
+   InPersistentCongestion(lost_packets):
      pto = smoothed_rtt + max(4 * rttvar, kGranularity) +
        max_ack_delay
      congestion_period = pto * kPersistentCongestionThreshold
      // Determine if all packets in the time period before the
      // newest lost packet, including the edges, are marked
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
     // largest newly lost packet, including the edges, are marked
```

File: draft-ietf-quic-recovery.md [hunk 8]
```diff
@@ -1434,26 +1411,27 @@ increases bytes_in_flight.
 ## On Packet Acknowledgement
 
 Invoked from loss detection's OnPacketAcked and is supplied with the
-acked_packet from sent_packets.
+newly acked_packets from sent_packets.
 
 ~~~
-   InCongestionRecovery(sent_time):
+   InLossRecovery(sent_time):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
   InCongestionRecovery(sent_time):
```
Revert this change. Was this deliberate? The term congestion recovery is used earlier in the spec and the term is defined.

File: draft-ietf-quic-recovery.md [hunk 9]
```diff
@@ -1434,26 +1411,27 @@ increases bytes_in_flight.
...
-       return
-     if (congestion_window < ssthresh):
-       // Slow start.
-       congestion_window += acked_packet.size
-     else:
+   OnPacketsAcked(acked_packets):
+     for (packet in acked_packets):
+       // Remove from bytes_in_flight.
+       bytes_in_flight -= packet.size
+       if (InLossRecovery(packet.time_sent)):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
       if (InCongestionRecovery(packet.time_sent)):
```

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: If a timeout occurred, then lost_packets should never be empty I believe?  I could add an assert for that if you'd like?

File: draft-ietf-quic-recovery.md (hunk 8 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: This was deliberate.  I've always heard it called loss recovery or just 'the recovery period'

But I can revert it.

File: draft-ietf-quic-recovery.md (hunk 7 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: I'm not sure largest makes this less clearer, at least to me.

File: draft-ietf-quic-recovery.md (hunk 2 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: I'm not a big fan of largest here and above.  If you want to make the change, maybe another PR, since this one is already quite large.

**Comment by ianswett** [comment]
@marten-seemann my reasoning was I thought it was critical to determine what packets were newly acked before invoking loss detection.  In some cases it may be(depending upon details of one's loss detection), but since we update largest_acked_packet[pn_space] outside the code which detects newly acked packets, it's not critical to the pseudocode as it stands today.

That being said, I think the parallelism is nice, so I'd prefer to keep it.

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:1250]: Hm, that's true. I was thinking about stream cancellations, but that doesn't matter here. assert SGTM

**Comment by janaiyengar** [comment]
@marten-seemann : It does change the order -- it does CC increase action after the decrease

**Comment by ianswett** [comment]
Assert added, merging.

File: draft-ietf-quic-recovery.md [hunk 10]
```diff
@@ -1108,14 +1108,18 @@ OnAckReceived(ack, pn_space):
     largest_acked_packet[pn_space] =
         max(largest_acked_packet[pn_space], ack.largest_acked)
 
+  // DetectNewlyAckedPackets finds packets that are newly
```
Comment (kazu-yamamoto) [inline draft-ietf-quic-recovery.md:1111]: `DetectNewlyAckedPackets` is not defined in `master`.
